bigtable-hbase-mirroring-client-1.x-parent/bigtable-hbase-mirroring-client-1.x/src/main/java/com/google/cloud/bigtable/mirroring/hbase1_x/bufferedmutator/SequentialMirroringBufferedMutator.java

```diff
@@ -0,0 +1,527 @@
+      this.failedPrimaryOperations.add(e.getRow(i));
+    }
+    this.userListener.onException(e, this);
+  }
+
+  @Override
+  protected void handleSecondaryException(RetriesExhaustedWithDetailsException e) {
+    reportWriteErrors(e);
+  }
+
+  @Override
+  protected void flushBufferedMutatorBeforeClosing()
+      throws ExecutionException, InterruptedException {
+    scheduleFlushAll().secondaryFlushFinished.get();
+  }
+
+  @Override
+  protected FlushFutures scheduleFlushScoped(
+      final List<Entry> dataToFlush, final FlushFutures previousFlushFutures) {
+    final SettableFuture<Void> secondaryFlushFinished = SettableFuture.create();
+
+    final ListenableFuture<Void> primaryFlushFinished =
+        schedulePrimaryFlush(previousFlushFutures.primaryFlushFinished);
+
+    final SettableFuture<Void> primaryFlushErrorsReported = SettableFuture.create();
+
+    Futures.addCallback(
+        primaryFlushFinished,
+        this.mirroringTracer.spanFactory.wrapWithCurrentSpan(
+            new FutureCallback<Void>() {
+              @Override
+              public void onSuccess(@NullableDecl Void aVoid) {
+                primaryFlushErrorsReported.set(null);
+                performSecondaryFlush(
+                    dataToFlush,
+                    secondaryFlushFinished,
+                    previousFlushFutures.secondaryFlushFinished);
+              }
+
+              @Override
+              public void onFailure(Throwable throwable) {
+                if (throwable instanceof RetriesExhaustedWithDetailsException) {
+                  // If user-defined listener has thrown an exception
+                  // (RetriesExhaustedWithDetailsException is the only exception that can be
+                  // thrown), we know that some of the writes failed. Our handler has already
+                  // handled those errors. We should also rethrow this exception when user
+                  // calls mutate/flush the next time.
+                  exceptionsToBeReportedToTheUser.addRetriesExhaustedException(
+                      (RetriesExhaustedWithDetailsException) throwable);
+                  primaryFlushErrorsReported.set(null);
+
+                  performSecondaryFlush(
+                      dataToFlush,
+                      secondaryFlushFinished,
+                      previousFlushFutures.secondaryFlushFinished);
+                } else {
+                  // In other cases, we do not know what caused the error and we have no idea
+                  // what was really written to the primary DB. We will behave as if nothing was
+                  // written and throw the exception to the user. Writing mutations to the faillog
+                  // would cause confusion as the user would think that those writes were successful
+                  // on primary, but they were not.
+                  exceptionsToBeReportedToTheUser.addThrowable(throwable);
+                  primaryFlushErrorsReported.set(null);
```

Comment:
not entirely clear to me what is designated for primary flushfinished and what is designtated for exceptionsToBeReportedToTheUser.